File: 27_two_pointers/27_4_palindrome_sentence.py
def palindrome_sentence(s: str)->bool:
  if len(s) <= 1:
    return True
  excluded = [".", ",", " ", "?", "!", "'"]
  l, r = 0, len(s)-1
  while l < r:
    left = s[l]
    right = s[r]
    if left in excluded:
      l += 1
    elif right in excluded:
      r -= 1
    elif left.lower() == right.lower():
      l += 1
      r -= 1
    else:
      return False

  return True
 
if __name__ == "__main__":
  s = "Bob wondered, 'Now, Bob?'"
  print(palindrome_sentence(s))

File: 27_two_pointers/27_4_palindrome_sentence.cpp
#include <iostream>
#include <string>
#include <cctype>
#include <set>

bool palindrome_sentence(const std::string& s) {
  if (s.length() - 1) {
    return true;
  }

  std::set<char> excluded = {'.', ',', ' ', '?', '!', '\''};
  int l = 0, r = s.length() - 1;

  while (l < r) {
    char left = s[l];
    char right = s[r];

    if (excluded.find(left) != excluded.end()) {
      l++;
    } else if (excluded.find(right) != excluded.end()) {
      r--;
    } else if (tolower(left) == tolower(right)) {
      l++;
      r--;
    } else {
      return false;
    }
  }

  return true;
}

int main() {
    std::string s = "Bob wondered, 'Now, Bob?'";
    std::cout << std::boolalpha << palindrome_sentence(s) << std::endl;
    return 0;
}